README.md
# files-to-consul-kv

Simple utility for bulk loading sets of [Consul key-value](https://www.consul.io/docs/agent/kv.html) entries via the [transactions API](https://www.consul.io/api/txn.html) where the source of those values exist on disk in a directory structure. 

For example given a simple directory structure like:
```
cd mykvs/

$ find . -print
.
./sub
./sub/key2
./key1

$ cat key1 
val1

$ cat sub/key2 
val2
```

You could use `fs2consulkv.py` to set all these in Consul under some root path:
```
 ./fs2consulkv.py \
    --fs-kv-path ./mykvs \
    --consul-url https://[consul-fqdn][:port] \
    --consul-acl-token xxxxxxx \
    --consul-data-center optional-dc \
    --consul-kv-root some/root/path/
 ```

Would result in your KVs in consul at:
```
https://[consul-fqdn][:port]/ui/mydc/kv/some/root/path/key1 = val1
https://[consul-fqdn][:port]/ui/mydc/kv/some/root/path/sub/key2 = val2
```

## Docker

Run via Docker:
https://hub.docker.com/r/bitsofinfo/files-to-consul-kv

```
docker run -i -v `pwd`/mykvs:/kvsource \
   bitsofinfo/files-to-consul-kv fs2consulkv.py \
   --fs-kv-path /kvsource \
   --consul-url https://[consul-fqdn][:port] \
   --consul-acl-token xxxxxxx \
   --consul-data-center optional-dc \
   --consul-kv-root some/root/path/
```

## Usage

```
$ ./fs2consulkv.py  -h

usage: fs2consulkv.py [-h] [-p FS_KV_PATH] [-k CONSUL_KV_ROOT]
                      [-z CONSUL_KV_ROOT_FILE] [-c CONSUL_URL]
                      [-t CONSUL_ACL_TOKEN] [-f CONSUL_ACL_TOKEN_FILE]
                      [-d CONSUL_DATA_CENTER] [-x] [-n] [-l LOG_LEVEL]
                      [-b LOG_FILE]

optional arguments:
  -h, --help            show this help message and exit
  -p FS_KV_PATH, --fs-kv-path FS_KV_PATH
                        Full or relative path to filesystem directory
                        containing the KV structure to send to consul
                        (default: ./)
  -k CONSUL_KV_ROOT, --consul-kv-root CONSUL_KV_ROOT
                        Root path in Consul KV by which all new keys will be
                        set, required. i.e. 'some/root/path' (default: None)
  -z CONSUL_KV_ROOT_FILE, --consul-kv-root-file CONSUL_KV_ROOT_FILE
                        Path to a file that contains the consul-kv-root
                        argument value, optional, can be used instead of
                        --consul-kv-root i.e. /path/to/consul-kv-root.txt
                        where the file contents contains the value
                        'some/root/path' (default: None)
  -c CONSUL_URL, --consul-url CONSUL_URL
                        Consul url, required. i.e. http[s]://[fqdn][:port]
                        (default: None)
  -t CONSUL_ACL_TOKEN, --consul-acl-token CONSUL_ACL_TOKEN
                        Consul acl token, required (default: None)
  -f CONSUL_ACL_TOKEN_FILE, --consul-acl-token-file CONSUL_ACL_TOKEN_FILE
                        Consul acl token file, path to a file that contains
                        the token value, required (default: None)
  -d CONSUL_DATA_CENTER, --consul-data-center CONSUL_DATA_CENTER
                        Consul data-center, optional. (default: None)
  -x, --skip-prompt     Skip confirmation and prompt (default: False)
  -n, --retain-trailing-newlines
                        Retain trailing newline chars (\n) in values files and
                        do not strip them. Default behavior is to strip them
                        (default: False)
  -s SLEEP_DELAY, --sleep-delay SLEEP_DELAY
                        Delay [in seconds] in kv upload loop, to avoid
                        overwhelming the consul server. Default behavior is
                        0.000 seconds (default: 0)
  -u CHUNK_SIZE, --chunk-size CHUNK_SIZE
                        Number of KV pairs uploaded at once. Default is 64,
                        the maximum allowed. (default: 64)
  -l LOG_LEVEL, --log-level LOG_LEVEL
                        log level, DEBUG, INFO, etc (default: DEBUG)
  -b LOG_FILE, --log-file LOG_FILE
                        Path to log file; default None = STDOUT (default:
                        None)
```


CHANGELOG.md
# 1.0.6

* added sleep/chunk size options: https://github.com/bitsofinfo/files-to-consul-kv/pull/5
 
# 1.0.5

* [#4 fix possible targetkv name mangling](https://github.com/bitsofinfo/files-to-consul-kv/pull/4)
* add license
# 1.0.4

exit non-zero on KV push failures

# 1.0.3

added `--consul-kv-root-file`

# 1.0.2

added `--consul-acl-token-file`

# 1.0.1

fix trimming kv root whitespace

# 1.0.0

initial release


fs2consulkv.py
#!/usr/bin/env python3

# https://www.consul.io/api/txn.html
# https://github.com/hashicorp/consul/issues/7278

import os
import base64
import requests
import json
import logging
import sys
import time
import argparse


# Yield successive n-sized 
# chunks from l. 
# https://www.geeksforgeeks.org/break-list-chunks-size-n-python/
def divide_chunks(l, n): 
    
    # looping till length l 
    for i in range(0, len(l), n):  
        yield l[i:i + n] 


def main():
    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)

    parser.add_argument('-p', '--fs-kv-path', dest='fs_kv_path', default="./", \
        help="Full or relative path to filesystem directory containing the KV structure to send to consul")

    parser.add_argument('-k', '--consul-kv-root', dest='consul_kv_root', default=None, \
        help="Root path in Consul KV by which all new keys will be set, required. i.e. 'some/root/path'")

    parser.add_argument('-z', '--consul-kv-root-file', dest='consul_kv_root_file', default=None, \
        help="Path to a file that contains the consul-kv-root argument value, optional, can be used instead of --consul-kv-root " + \
            "i.e. /path/to/consul-kv-root.txt where the file contents contains the value 'some/root/path'")

    parser.add_argument('-c', '--consul-url', dest='consul_url', default=None, \
        help="Consul url, required. i.e. http[s]://[fqdn][:port]")

    parser.add_argument('-t', '--consul-acl-token', dest='consul_acl_token', default=None, \
        help="Consul acl token, required")

    parser.add_argument('-f', '--consul-acl-token-file', dest='consul_acl_token_file', default=None, \
        help="Consul acl token file, path to a file that contains the token value, required")

    parser.add_argument('-d', '--consul-data-center', dest='consul_data_center', default=None, \
        help="Consul data-center, optional.")

    parser.add_argument('-x', '--skip-prompt', action='store_true', default=False, \
        help="Skip confirmation and prompt")

    parser.add_argument('-n', '--retain-trailing-newlines', action='store_true', default=False, \
        help="Retain trailing newline chars (\\n) in values files and do not strip them. Default behavior is to strip them")

    parser.add_argument('-s', '--sleep-delay', dest='sleep_delay', default=0, \
        help="Delay [in seconds] in kv upload loop, to avoid overwhelming the consul server. Default behavior is 0.000 seconds")
    parser.add_argument('-u', '--chunk-size', dest='chunk_size', default=64, \
        help="Number of KV pairs uploaded at once. Default is 64, the maximum allowed.")

    parser.add_argument('-l', '--log-level', dest='log_level', default="DEBUG", \
        help="log level, DEBUG, INFO, etc")
    parser.add_argument('-b', '--log-file', dest='log_file', default=None, \
        help="Path to log file; default None = STDOUT")

    args = parser.parse_args()

    dump_help = False

    if not args.consul_kv_root and not args.consul_kv_root_file:
        dump_help = True
   
    if dump_help:
        parser.print_help()
        sys.exit(1)

    logging.basicConfig(level=logging.getLevelName(args.log_level),
                        format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
                        filename=args.log_file,filemode='w')
    logging.Formatter.converter = time.gmtime

    url = "{}/v1/txn".format(args.consul_url)

    if args.consul_data_center is not None:
        url += "?dc={}".format(args.consul_data_center)

    consul_acl_token = args.consul_acl_token
    if args.consul_acl_token_file:
        with open (args.consul_acl_token_file, "r") as tokenfile:
            value = tokenfile.read()
            if value and value.strip() == '':
                logging.error("--consul-acl-token-file {} contains nothing!".format(args.consul_acl_token_file))
                sys.exit(1)
            else:
                consul_acl_token = value

    if not consul_acl_token or consul_acl_token.strip() == '':
        logging.error("--consul-acl-token-file or --consul-acl-token is required!")
        sys.exit(1)


    consul_kv_root = args.consul_kv_root
    if args.consul_kv_root_file:
        with open (args.consul_kv_root_file, "r") as f:
            consul_kv_root = f.read()
        logging.debug("sourced --consul-kv-root-file at {} with value {}".format(args.consul_kv_root_file,consul_kv_root))


    headers = {
        'Content-Type': "application/json",
        'X-Consul-Token': consul_acl_token,
        'User-Agent': "github.com/bitsofinfo/files-to-consul-kv",
        'Accept': "*/*",
        'Cache-Control': "no-cache"
    }

    try:
        kvs = []

        if consul_kv_root.strip().endswith('/'):
            consul_kv_root = consul_kv_root.strip()[:-1]

        for root, dirs, files in os.walk(args.fs_kv_path):

            for name in files:
                filepath = os.path.join(root, name)
            
                targetkv = filepath.replace(args.fs_kv_path, "", 1)

                with open (filepath, "r") as kvfile:
                    value = kvfile.read()

                if not args.retain_trailing_newlines:
                     if value.endswith('\n'):
                        value = value[:-1]

                if targetkv.startswith('/'):
                    targetkv = targetkv[1:]

                kvs.append({
                    "KV": {
                    "Verb": "set",
                    "Key": "{}/{}".format(consul_kv_root,targetkv),
                    "Value": "{}".format(base64.b64encode(value.encode("utf-8")).decode("utf-8"))
                    }
                })

        if not args.skip_prompt:
            print(json.dumps(kvs,indent=2))
            proceed = input("\n\nYou are about to 'set' the above consul KV paths:\n... that were consumed from {}\n... will be PUT against {}\n... and set relative from {}/\n\nAre you sure?: (y|n):".format(args.fs_kv_path,url,consul_kv_root)).strip()
            if proceed.lower() != 'y':
                logging.info("Exiting, confirmation prompt input was: " + proceed)
                sys.exit(1)


        # we can only max send 64 per request...
        # https://github.com/hashicorp/consul/issues/7278
        kv_chunks = list(divide_chunks(kvs, int(args.chunk_size))) 

        logging.info("Number of kvs totals: {}, this has to be split up " \
            "into {} {} kv chunks: https://github.com/hashicorp/consul/issues/7278".format(len(kvs),len(kv_chunks),int(args.chunk_size)))
        
        exit_with_exit_code = 0

        for kvchunk in kv_chunks:

            logging.info("PUTing chunk with {} keys @ {}".format(len(kvchunk),url))

            response = requests.request("PUT", url, data=json.dumps(kvchunk), headers=headers)

            time.sleep(float(args.sleep_delay))
            
            if response.status_code == 200:
                logging.debug("KVs 'set' OK: {}".format(response.content))
            else:
                logging.error("KVs 'set' http response code: {} FAILED: {} ".format(response.status_code,json.loads(response.content)))
                exit_with_exit_code = 1

        sys.exit(exit_with_exit_code)

    except Exception as e:
        logging.exception("error PUTing consul KVs via /txn: POST-DATA={} ERROR={} " \
            .format(kvs,str(sys.exc_info()[:2])))

if __name__ == '__main__':
    main()
